quil-rs/src/program/mod.rs

```diff
@@ -250,43 +251,40 @@ impl Program {
         Ok(expanded_program)
     }
 
-    pub fn to_instructions(&self, include_headers: bool) -> Vec<Instruction> {
-        let mut result = vec![];
+    pub fn to_instructions(&self) -> Vec<Instruction> {
```

Comment:
`to_` is actually the correct prefix here, according to the [Rust API naming guidelines](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv)

(Summary: `as_` == cheap borrowed, `to_` == expensive operation, likely with cloning, `into_` == owned-to-owned conversion)

Reply:
I do like the idea of a method returning an iterator. Per convention, it would probably just be called `instructions()`.

Reply:
For my own learning, I tried implementing an `instructions()` method that returns an iterator but ran into trouble finding a way to do it without cloning. As a slimmed down example, consider:

```rust
    pub fn instructions(&'a self) -> impl Iterator<Item = &Instruction> + 'a {
        self.memory_regions
            .iter()
            .map(|(name, descriptor)| {
                Instruction::Declaration(Declaration {
                    name: *name,
                    size: descriptor.size,
                    sharing: descriptor.sharing,
                })
            }) // => Returns an Iterator with `Item = Instruction`
            .chain(self.instructions.iter()) // Returns an Iterator with `Item = &Instruction`
    }
```

Because the types of the iterators are mismatched, Rust won't compile this code. To avoid cloning `self.instructions`, we have to create an iterator of references to the Declaration instructions, which can't be done with out collecting them into something like a vector first. We'd have to do the same for `frames` and `calibrations` as well, since we also build new `Instruction`s out of their data. Maybe there is a workaround for this, but I couldn't find one after playing with it for a bit.